baseband/guppi/header.py

```diff
@@ -0,0 +1,379 @@
+
+    or, to also keep the comments::
+
+        GUPPIHeader(((key, (header[key], header.comments[key]))
+                      for key in header))
+    """
+
+    _properties = ('payload_nbytes', 'frame_nbytes', 'bps', 'nchan', 'npol',
+                   'sample_shape', 'sample_rate', 'sideband', 'overlap',
+                   'samples_per_frame', 'offset', 'start_time',
+                   'time')
+    """Properties accessible/usable in initialisation for all headers."""
+
+    _defaults = [('BACKEND', 'GUPPI'),
+                 ('BLOCSIZE', 0),
+                 ('STT_OFFS', 0),
+                 ('PKTIDX', 0),
+                 ('OVERLAP', 0),
+                 ('SRC_NAME', 'unset'),
+                 ('TELESCOP', 'unset'),
+                 ('PKTFMT', '1SFA'),
+                 ('NBITS', 8),
+                 ('NPOL', 1),
+                 ('OBSNCHAN', 1)]
+
+    def __init__(self, *args, **kwargs):
+        verify = kwargs.pop('verify', True)
+        mutable = kwargs.pop('mutable', True)
+        # Comments handled by fits.Header__init__().
+        super(GUPPIHeader, self).__init__(*args, **kwargs)
+        self.mutable = mutable
+        if verify:
+            self.verify()
+
+    def verify(self):
+        """Basic check of integrity."""
+        assert all(key in self for key in ('BLOCSIZE',
+                                           'PKTIDX'))
+        assert self['PKTFMT'] == '1SFA'
+
+    def copy(self):
+        """Create a mutable and independent copy of the header."""
+        newfitsheader = super(GUPPIHeader, self).copy()
+        return self.__class__(newfitsheader)
+
+    def __copy__(self):
+        return self.copy()
+
+    @classmethod
+    def fromfile(cls, fh, verify=True):
+        """
+        Reads in GUPPI header block from a file.
+
+        Parameters
+        ----------
+        fh : filehandle
+            To read data from.
+        verify: bool, optional
+            Whether to do basic checks on whether the header is valid.
+            Default: `True`.
+        """
+        header_start = fh.tell()
+        # Find the size of the header.  GUPPI header entries are 80 char long
+        # with <=8 char keyword names.  "=" is always the 9th char.
+        line = '========='
+        while line[8] in ('=', ' '):
+            line = fh.read(80).decode('ascii')
+            if line[:3] == 'END':
+                break
+
+        header_end = fh.tell()
+        fh.seek(header_start)
+        hdr = fh.read(header_end - header_start).decode('ascii')
+        # Calls fits.Header to read header.
+        fits_header = super(GUPPIHeader, cls).fromstring(hdr)
+        return cls(fits_header, verify=verify, mutable=False)
+
+    def tofile(self, fh):
+        """Write GUPPI file header to filehandle.
+
+        Uses `~astropy.io.fits.Header.tostring`.
+        """
+        fh.write(self.tostring(padding=False).encode())
+        #raise NotImplementedError('GUPPI format does not have write support.')
```

Comment:
Obviously we should either remove the comment or decide on sticking with not having `write` support...